### DISCUSSION TITLE
RGBA color support inside a Jupyter Notebook

### DISCUSSION THREAD

**Discussion by GuillaumeFavelier** [pyvista/pyvista#2642] [Q&A]
When I use **RGBA** colors inside a Jupyter Notebook, I obtain an unexpected result:

```py
import pyvista as pv
import numpy as np

sphere = pv.Sphere()
scalars = np.ones((sphere.GetNumberOfPoints(), 4)) * 255.0
scalars = scalars.astype('ubyte')

p = pv.Plotter()
p.add_mesh(sphere, rgba=True, scalars=scalars)
p.show()
```

![image](https://user-images.githubusercontent.com/18143289/59596130-68bc3f00-90f7-11e9-985d-2f9c87825ff2.png)

When I run this script natively, I get the right result of course:

![image](https://user-images.githubusercontent.com/18143289/59596179-812c5980-90f7-11e9-8228-3484e7600dfa.png)

**Comment by GuillaumeFavelier** [comment]
Now that I think about it, maybe this could be fixed by [#258](https://github.com/pyvista/pyvista/issues/258) ? Well, it's also possible that my code snippet isn't correct.

**Comment by banesullivan** [comment]
It's not your code, it's `panel`'s converter to VTKjs (see https://github.com/pyviz/panel/issues/408).

At the moment, we are using `panel`'s export script rather than ours to convert the scene to VTKjs when making the embeddable rendering scenes; there are a number of bugs (and most of those bugs exist in our script too). We need to move forward with bug fixing the export script in PyVista to work with `panel` to avoid this issue.

Perhaps https://github.com/pyvista/pyvista/issues/258 would be a better approach and more worth our time. K3D seems far more robust than VTKjs so maybe we should focus on creating a way to display PyVista scenes with K3D rather than fixing the VTKjs script.

**Comment by banesullivan** [comment]
Also, I've found a hack of a fix. If you only have a single array present in the dataset you'd like to send to VTKjs, it'll work:

```py
import pyvista as pv
import numpy as np

sphere = pv.Sphere()
scalars = np.ones((sphere.GetNumberOfPoints(), 4)) * 255.0
scalars = scalars.astype('ubyte')

# ONLY have one array present for VTKjs
# del sphere.point_arrays['Normals']
sphere.clear_arrays()

p = pv.Plotter()
p.add_mesh(sphere, rgba=True, scalars=scalars)
p.show()
```

<img width="907" alt="Screen Shot 2019-06-17 at 8 30 12 AM" src="https://user-images.githubusercontent.com/22067021/59612333-23d5df80-90da-11e9-9739-a5d287126c39.png">

**Comment by GuillaumeFavelier** [comment]
This hack is already amazing as a short-term solution, thank you so much! I will try it. So out of curiosity, is it the array of `Normals` which is displayed by default then?

And the choice between K3D and VTKjs looks like a dilemma. Even if I looked their showcase gallery, I still need to try K3D myself to have a better understanding of its capabilities.

**Comment by banesullivan** [comment]
> So out of curiosity, is it the array of Normals which is displayed by default then?

The VTKjs export script has trouble choosing the active array. It always grabs the array at the 0th index so if you remove all other arrays other than what you want to plot it should work fine.
 
>  I still need to try K3D myself to have a better understanding of its capabilities.

I definitely need to look into it more as well
